zerver/tornado/handlers.py

```diff
@@ -180,175 +133,112 @@ def delete(self, *args: Any, **kwargs: Any) -> None:
-            signals.got_request_exception.send(sender=self.__class__, request=request)
-            response = self.handle_uncaught_exception(request, resolver, sys.exc_info())
-
-        return response
-
-class AsyncDjangoHandler(AsyncDjangoHandlerBase):
-    def zulip_finish(self, response: Dict[str, Any], request: HttpRequest,
+    def zulip_finish(self, result_dict: Dict[str, Any], request: HttpRequest,
                      apply_markdown: bool) -> None:
-        # Make sure that Markdown rendering really happened, if requested.
-        # This is a security issue because it's where we escape HTML.
-        # c.f. ticket #64
-        #
-        # apply_markdown=True is the fail-safe default.
-        if response['result'] == 'success' and 'messages' in response and apply_markdown:
-            for msg in response['messages']:
+        # Function called when we want to break a long-polled
+        # get_events request and return a response to the client.
+        if result_dict['result'] == 'success' and 'messages' in result_dict and apply_markdown:
+            for msg in result_dict['messages']:
                 if msg['content_type'] != 'text/html':
                     self.set_status(500)
                     self.finish('Internal error: bad message format')
-        if response['result'] == 'error':
+        if result_dict['result'] == 'error':
             self.set_status(400)
 
-        # Call the Django response middleware on our object so that
-        # e.g. our own logging code can run; but don't actually use
-        # the headers from that since sending those to Tornado seems
-        # tricky; instead just send the (already json-rendered)
-        # content on to Tornado
-        django_response = json_response(res_type=response['result'],
-                                        data=response, status=self.get_status())
-        django_response = self.apply_response_middleware(request, django_response,
-                                                         request._resolver)
+        # The `result` dictionary contains the data we want to return
+        # to the client.  We want to do so in a proper HTTP response
+        # after running our Django response middleware (which does
+        # things like log the request, add rate-limit headers, etc.).
+        # Making an HttpResponse object with json_response is easy.
+        # What's tricky is getting Django to run process_response
+        # middleware on it without writing a lot of duplicate code.
+        #
+        # Django runs the process_response middleware functions inside
+        # MiddlewareMixin.__call__, which also runs the
+        # process_request middleware.  That code path is activated
+        # inside Django BaseHandler's `self.get_response` and
+        # `load_middleware` functions, which effectively end up doing:
+        #    response = m1(m2(m3(...(self._get_response(request)))))
```

Comment:
One bit of context that wasn't clear to me, until I'd looked back and forth a few times between this explanation and the `_get_response` and `middleware_mixin__call___hack` code below, is that this all happens inside Django's `get_response`. (Just now, as I'm writing this comment, I see that this paragraph actually mentions `get_response`; I was only seeing `_get_response`.)

One thing that might help would be to expand "which also runs" and "is activated inside" into a few more lines of pseudocode akin to this line here -- I think this line is quite helpful already.